### PR TITLE
[CDAP-901] Feature/namespace client

### PR DESCRIPTION
**Pull request by bdmogal**
Added client for namespace admin API

### CODE REVIEW

**Comment by bdmogal** [comment]
Build: https://builds.cask.co/browse/CDAP-DUT402

File: cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import javax.inject.Inject;
+
+/**
+ * Client to interact with CDAP namespaces
+ */
+public class NamespaceClient {
+  private final RESTClient restClient;
+  private final ClientConfig config;
+
+  private static final String NAMESPACE_ENTITY_TYPE = "namespace";
```
Comment (ParthGandhi) [inline cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java]: Better to have static final fields on top.

File: cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    CannotBeDeletedException {
+    HttpResponse response = restClient.execute(HttpMethod.DELETE,
+                                               config.resolveURL(String.format("namespaces/%s", namespaceId)),
+                                               config.getAccessToken(),
+                                               HttpURLConnection.HTTP_NOT_FOUND,
+                                               HttpURLConnection.HTTP_FORBIDDEN);
+    if (HttpURLConnection.HTTP_NOT_FOUND == response.getResponseCode()) {
+      throw new NotFoundException(NAMESPACE_ENTITY_TYPE, namespaceId);
+    }
+    if (HttpURLConnection.HTTP_FORBIDDEN == response.getResponseCode()) {
```
Comment (ParthGandhi) [inline cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java]: `else-if`

File: cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    if (displayName != null) {
+      builder.setDisplayName(displayName);
+    }
+    if (description != null) {
+      builder.setDescription(description);
+    }
+    String body = new Gson().toJson(builder.build());
+    HttpRequest request = HttpRequest.put(url).withBody(body).build();
+    HttpResponse response = restClient.upload(request, config.getAccessToken(), HttpURLConnection.HTTP_BAD_REQUEST);
+    String responseBody = response.getResponseBodyAsString();
```
Comment (ParthGandhi) [inline cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java]: Better to specify the `Charset` here.

File: cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    if (response.getResponseCode() == HttpURLConnection.HTTP_BAD_REQUEST) {
+      throw new BadRequestException("Bad request: " + responseBody);
+    }
+    if (responseBody != null && responseBody.contains("already exists")) {
```
Comment (ParthGandhi) [inline cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java]: This seems extremely fickle. Is there no better way?

File: cdap-client-tests/src/test/java/co/cask/cdap/client/NamespaceClientTestRun.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    Assert.assertEquals(3, namespaces.size());
+    meta = namespaceClient.get(NO_DESCRIPTION);
+    Assert.assertEquals(NO_DESCRIPTION, meta.getId());
+    Assert.assertEquals(TEST_DISPLAY_NAME, meta.getDisplayName());
+    Assert.assertEquals("", meta.getDescription());
+
+    // cleanup
+    namespaceClient.delete(TEST_NAMESPACE_ID);
+    namespaceClient.delete(NO_DISPLAY_NAME);
+    namespaceClient.delete(NO_DESCRIPTION);
```
Comment (ParthGandhi) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/NamespaceClientTestRun.java]: Good to also assert that `namespaceClient.list()` now returns an empty list.

File: cdap-client-tests/src/test/java/co/cask/cdap/client/NamespaceClientTestRun.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    // verify that the namespace got created correctly
+    namespaces = namespaceClient.list();
+    Assert.assertEquals(1, namespaces.size());
+    NamespaceMeta meta = namespaceClient.get(TEST_NAMESPACE_ID);
+    Assert.assertEquals(TEST_NAMESPACE_ID, meta.getId());
+    Assert.assertEquals(TEST_DISPLAY_NAME, meta.getDisplayName());
+    Assert.assertEquals(TEST_DESCRIPTION, meta.getDescription());
+
+    // create and verify namespace without displayName
+    builder = new NamespaceMeta.Builder();
```
Comment (ParthGandhi) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/NamespaceClientTestRun.java]: Any reason you can't verify no `displayName` and `description` as part of the same test block?

File: cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+  private static final String NAMESPACE_ENTITY_TYPE = "namespace";
+
+  @Inject
+  public NamespaceClient(ClientConfig config) {
+    this.config = config;
+    this.restClient = RESTClient.create(config);
+  }
+
+  /**
+   * Lists all namespaces in CDAP.
```
Comment (ParthGandhi) [inline cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java]: I *could* be wrong about this, but I think it is incorrect to say "namespace in CDAP". I think the convention is "Lists all namespaces in a CDAP instance" or just "Lists all namespaces".

**Comment by ParthGandhi** [comment]
Couple of minor comments.

File: cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java (hunk 4 above)
Comment (bdmogal) [inline cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java]: There is no better way other than string matching unfortunately. The create API is a PUT API. So we wanted to make it idempotent. Hence even if the namespace already exists, it returns a 200 with a message that the namespace already exists in the response body. 

However, I agree that we can make this check a little stricter. Have added an ``equals`` check as opposed to the current ``contains`` check. Let me know what you think.

File: cdap-client-tests/src/test/java/co/cask/cdap/client/NamespaceClientTestRun.java (hunk 5 above)
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/NamespaceClientTestRun.java]: Good point. Fixed.

File: cdap-client-tests/src/test/java/co/cask/cdap/client/NamespaceClientTestRun.java (hunk 6 above)
Comment (bdmogal) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/NamespaceClientTestRun.java]: Absolutely no reason. Fixed.

File: cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    NamespaceMeta.Builder builder = new NamespaceMeta.Builder();
+    String displayName = namespaceMeta.getDisplayName();
+    String description = namespaceMeta.getDescription();
+    if (displayName != null) {
+      builder.setDisplayName(displayName);
+    }
+    if (description != null) {
+      builder.setDescription(description);
+    }
+    String body = new Gson().toJson(builder.build());
```
Comment (jwang47) [inline cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java]: Usually we put `private static final Gson GSON = new Gson();` whenever we have a class that uses gson.

**Comment by jwang47** [comment]
Just one minor comment from me.

File: cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java (hunk 8 above)
Comment (bdmogal) [inline cdap-client/src/main/java/co/cask/cdap/client/NamespaceClient.java]: Fixed.

**Comment by bdmogal** [comment]
Addressed comments. Build passing - https://builds.cask.co/browse/CDAP-DUT402-2.

**Comment by jwang47** [comment]
LGTM

**Comment by bdmogal** [comment]
@ParthGandhi , merging this for now since this has a dependent PR #939. If you have any more comments about this, I'll take care of them in a separate PR.

**Pull request merged**